packages/block-editor/src/components/block-toolbar-link-control/index.js

```diff
@@ -0,0 +1,88 @@
+/**
+ * WordPress dependencies
+ */
+import { ToolbarButton, ToolbarGroup } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+import { useCallback, useMemo, useState } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import BlockControls from '../block-controls';
+import SettingsToolbarItem from './settings-toolbar-item';
+import LinkInputToolbarItem from './link-input-toolbar-item';
+import computeNiceURL from './compute-nice-url';
+import ToolbarLinkControlContext from './context';
+
+export default function ToolbarLinkControl( {
+	initialLink,
+	createSuggestion,
+	close,
+	onChange,
+} ) {
+	const [ currentLink, setCurrentLink ] = useState( {
+		...initialLink,
+		url: computeNiceURL( initialLink.url ),
+	} );
+
+	const [ preferredDropdown, setPreferredDropdown ] = useState(
+		'suggestions'
+	);
+
+	const updateCurrentLink = useCallback(
+		( data, replace = false ) => {
+			setCurrentLink( {
+				...( replace ? {} : currentLink ),
+				...data,
+				url: data.url ? computeNiceURL( data.url ) : currentLink.url,
+			} );
+		},
+		[ currentLink ]
+	);
+
+	const finishLinkEditing = ( acceptChanges = true ) => {
+		if ( acceptChanges ) {
+			onChange( currentLink );
+		}
+		close();
+	};
+
+	const contextValue = useMemo(
+		() => ( {
+			createSuggestion,
+			currentLink,
+			updateCurrentLink,
+			preferredDropdown,
+			setPreferredDropdown,
+		} ),
+		[
+			createSuggestion,
+			currentLink,
+			updateCurrentLink,
+			preferredDropdown,
+			setPreferredDropdown,
+		]
+	);
+
+	return (
+		<BlockControls __experimentalIsExpanded={ true }>
+			<ToolbarLinkControlContext.Provider value={ contextValue }>
+				<ToolbarGroup>
+					<LinkInputToolbarItem />
+				</ToolbarGroup>
+				<ToolbarGroup className="toolbar-link-control__input-group">
+					<SettingsToolbarItem />
+				</ToolbarGroup>
+				<ToolbarGroup>
```

Comment:
Any reason for wrapping each element in its own group?

Reply:
It provides vertical dividers between different elements, it seems like it varies from design to design so maybe we don't need it.

Reply:
Semantically, toolbar groups should be used to separate different groups of buttons from each other. As far as I can tell, there's no reason to use multiple toolbar groups here.

Reply:
Actually, now that I think about it, maybe the "Done" button should stay separated from the others.